File: CS/BlockSelection/BlockSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraTreeList;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraTreeList.Nodes.Operations;
using DevExpress.XtraTreeList.Columns;
using DevExpress.XtraTreeList.Nodes;

namespace BlockSelection
{
    class BlockSelector
    {
        private TreeList fList;
        private Block fBlock;

        public BlockSelector(TreeList list)
        {
            if (list == null)
                throw new Exception("A valid TreeList instance must be passed to the constructor");

            fList = list;
            fBlock = new Block();

            fList.FocusedColumnChanged += new FocusedColumnChangedEventHandler(fList_FocusedColumnChanged);
            fList.FocusedNodeChanged += new FocusedNodeChangedEventHandler(fList_FocusedNodeChanged);
            fList.MouseMove += new System.Windows.Forms.MouseEventHandler(fList_MouseMove);
            fList.MouseDown += new MouseEventHandler(fList_MouseDown);
            fList.NodeCellStyle += new GetCustomNodeCellStyleEventHandler(fList_NodeCellStyle);
        }

        public Block Block { get { return fBlock; } }

        public TreeList List { get { return fList; } }

        private void fList_NodeCellStyle(object sender, GetCustomNodeCellStyleEventArgs e)
        {
            if (fBlock.Contains(e.Column.VisibleIndex, e.Node.Id))
                e.Appearance.BackColor = SystemColors.Highlight;
        }

        private delegate void SelectEntireNodeDlg(TreeListNode node);

        private void SelectEntireNode(TreeListNode node)
        {
            fBlock.X1 = 0;
            fBlock.X2 = fList.VisibleColumns.Count - 1;
            fBlock.Y2 = node.Id;
            if (fBlock.Modified && !fBlock.IsEmpty()) InvalidateBlock();
        }

        void fList_MouseDown(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left)
            {
                TreeListHitInfo hInfo = fList.CalcHitInfo(e.Location);
                if (hInfo.HitInfoType == HitInfoType.RowIndicator)
                {
                    fList.BeginInvoke(new SelectEntireNodeDlg(SelectEntireNode), hInfo.Node);
                }
            }
        }

        private void fList_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left)
            {
                TreeListHitInfo hInfo = fList.CalcHitInfo(e.Location);
                if (hInfo.HitInfoType == HitInfoType.Cell)
                    fBlock.X2 = hInfo.Column.VisibleIndex;
                if (hInfo.HitInfoType == HitInfoType.Cell || hInfo.HitInfoType == HitInfoType.RowIndicator)
                {
                    fBlock.Y2 = hInfo.Node.Id;
                    if (fBlock.Modified && !fBlock.IsEmpty()) InvalidateBlock();
                }
            }
        }

        public void InvalidateBlock() { fList.LayoutChanged(); }

        private void fList_FocusedNodeChanged(object sender, FocusedNodeChangedEventArgs e) { if (fList.FocusedColumn != null) UpdateBlock(); }

        private void UpdateBlock()
        {
            if (Control.ModifierKeys == Keys.Shift)
            {
                fBlock.X2 = fList.FocusedColumn.VisibleIndex;
                fBlock.Y2 = fList.FocusedNode.Id;
            }
            else
            {
                fBlock.X1 = fList.FocusedColumn.VisibleIndex;
                fBlock.Y1 = fList.FocusedNode.Id;
                fBlock.X2 = fBlock.X1;
                fBlock.Y2 = fBlock.Y1;
            }
            if (fBlock.Modified) InvalidateBlock();
        }

        void fList_FocusedColumnChanged(object sender, FocusedColumnChangedEventArgs e) { UpdateBlock(); }

        public string GetSelectedValues()
        {
            string result = string.Empty;
            if (!fBlock.IsEmpty())
            {
                SelectionOperation operation = new SelectionOperation(fBlock);
                fList.NodesIterator.DoOperation(operation);
                result = operation.Result;
            }
            return result;
        }
    }

    public class Block
    {
        private int FX1;
        private int FX2;
        private int FY1;
        private int FY2;
        private bool FModified;

        public bool Between(int a, int b, int c)
        {
            if (a > b)
            {
                int temp = a;
                a = b;
                b = temp;
            }
            return (a <= c) && (c <= b);
        }

        public bool Contains(int x, int y)
        {
            return Between(X1, X2, x) && Between(Y1, Y2, y);
        }
        public bool IsEmpty()
        {
            return (X1 == X2 && Y1 == Y2);
        }

        #region Coordinates
        public int X1
        {
            get
            {
                return FX1;
            }
            set
            {
                if (FX1 != value)
                {
                    FX1 = value;
                    FModified = true;
                }
            }
        }
        public int X2
        {
            get
            {
                return FX2;
            }
            set
            {
                if (FX2 != value)
                {
                    FX2 = value;
                    FModified = true;
                }
            }
        }
        public int Y1
        {
            get
            {
                return FY1;
            }
            set
            {
                if (FY1 != value)
                {
                    FY1 = value;
                    FModified = true;
                }
            }
        }
        public int Y2
        {
            get
            {
                return FY2;
            }
            set
            {
                if (FY2 != value)
                {
                    FY2 = value;
                    FModified = true;
                }
            }
        }
        #endregion

        public bool Modified
        {
            get
            {
                return FModified;
            }
        }
    }

    public class SelectionOperation : TreeListOperation
    {
        private Block block;
        private string result = string.Empty;
        private const string CellDelimeter = "\t";
        private const string LineDelimeter = "\r\n";

        public SelectionOperation(Block block) { this.block = block; }

        public string Result { get { return result; } }

        public override void Execute(DevExpress.XtraTreeList.Nodes.TreeListNode node)
        {
            if (!block.Between(block.Y1, block.Y2, node.Id)) return;
            foreach (TreeListColumn column in node.TreeList.Columns)
                if (block.Contains(column.VisibleIndex, node.Id))
                {
                    result += node.GetDisplayText(column);
                    result += CellDelimeter;
                }
            result += LineDelimeter;
        }
    }

}

File: VB/BlockSelection/BlockSelector.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraTreeList
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList.Nodes.Operations
Imports DevExpress.XtraTreeList.Columns
Imports DevExpress.XtraTreeList.Nodes

Namespace BlockSelection
	Friend Class BlockSelector
		Private fList As TreeList
		Private fBlock As Block

		Public Sub New(ByVal list As TreeList)
			If list Is Nothing Then
				Throw New Exception("A valid TreeList instance must be passed to the constructor")
			End If

			fList = list
			fBlock = New Block()

			AddHandler fList.FocusedColumnChanged, AddressOf fList_FocusedColumnChanged
			AddHandler fList.FocusedNodeChanged, AddressOf fList_FocusedNodeChanged
			AddHandler fList.MouseMove, AddressOf fList_MouseMove
			AddHandler fList.MouseDown, AddressOf fList_MouseDown
			AddHandler fList.NodeCellStyle, AddressOf fList_NodeCellStyle
		End Sub

		Public ReadOnly Property Block() As Block
			Get
				Return fBlock
			End Get
		End Property

		Public ReadOnly Property List() As TreeList
			Get
				Return fList
			End Get
		End Property

		Private Sub fList_NodeCellStyle(ByVal sender As Object, ByVal e As GetCustomNodeCellStyleEventArgs)
			If fBlock.Contains(e.Column.VisibleIndex, e.Node.Id) Then
				e.Appearance.BackColor = SystemColors.Highlight
			End If
		End Sub

		Private Delegate Sub SelectEntireNodeDlg(ByVal node As TreeListNode)

		Private Sub SelectEntireNode(ByVal node As TreeListNode)
			fBlock.X1 = 0
			fBlock.X2 = fList.VisibleColumns.Count - 1
			fBlock.Y2 = node.Id
			If fBlock.Modified AndAlso Not fBlock.IsEmpty() Then
				InvalidateBlock()
			End If
		End Sub

		Private Sub fList_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
			If e.Button = MouseButtons.Left Then
				Dim hInfo As TreeListHitInfo = fList.CalcHitInfo(e.Location)
				If hInfo.HitInfoType = HitInfoType.RowIndicator Then
					fList.BeginInvoke(New SelectEntireNodeDlg(AddressOf SelectEntireNode), hInfo.Node)
				End If
			End If
		End Sub

		Private Sub fList_MouseMove(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
			If e.Button = MouseButtons.Left Then
				Dim hInfo As TreeListHitInfo = fList.CalcHitInfo(e.Location)
				If hInfo.HitInfoType = HitInfoType.Cell Then
					fBlock.X2 = hInfo.Column.VisibleIndex
				End If
				If hInfo.HitInfoType = HitInfoType.Cell OrElse hInfo.HitInfoType = HitInfoType.RowIndicator Then
					fBlock.Y2 = hInfo.Node.Id
					If fBlock.Modified AndAlso Not fBlock.IsEmpty() Then
						InvalidateBlock()
					End If
				End If
			End If
		End Sub

		Public Sub InvalidateBlock()
			fList.LayoutChanged()
		End Sub

		Private Sub fList_FocusedNodeChanged(ByVal sender As Object, ByVal e As FocusedNodeChangedEventArgs)
			If fList.FocusedColumn IsNot Nothing Then
				UpdateBlock()
			End If
		End Sub

		Private Sub UpdateBlock()
			If Control.ModifierKeys = Keys.Shift Then
				fBlock.X2 = fList.FocusedColumn.VisibleIndex
				fBlock.Y2 = fList.FocusedNode.Id
			Else
				fBlock.X1 = fList.FocusedColumn.VisibleIndex
				fBlock.Y1 = fList.FocusedNode.Id
				fBlock.X2 = fBlock.X1
				fBlock.Y2 = fBlock.Y1
			End If
			If fBlock.Modified Then
				InvalidateBlock()
			End If
		End Sub

		Private Sub fList_FocusedColumnChanged(ByVal sender As Object, ByVal e As FocusedColumnChangedEventArgs)
			UpdateBlock()
		End Sub

		Public Function GetSelectedValues() As String
			Dim result As String = String.Empty
			If Not fBlock.IsEmpty() Then
				Dim operation As New SelectionOperation(fBlock)
				fList.NodesIterator.DoOperation(operation)
				result = operation.Result
			End If
			Return result
		End Function
	End Class

	Public Class Block
		Private FX1 As Integer
		Private FX2 As Integer
		Private FY1 As Integer
		Private FY2 As Integer
		Private FModified As Boolean

		Public Function Between(ByVal a As Integer, ByVal b As Integer, ByVal c As Integer) As Boolean
			If a > b Then
				Dim temp As Integer = a
				a = b
				b = temp
			End If
			Return (a <= c) AndAlso (c <= b)
		End Function

		Public Function Contains(ByVal x As Integer, ByVal y As Integer) As Boolean
			Return Between(X1, X2, x) AndAlso Between(Y1, Y2, y)
		End Function
		Public Function IsEmpty() As Boolean
			Return (X1 = X2 AndAlso Y1 = Y2)
		End Function

		#Region "Coordinates"
		Public Property X1() As Integer
			Get
				Return FX1
			End Get
			Set(ByVal value As Integer)
				If FX1 <> value Then
					FX1 = value
					FModified = True
				End If
			End Set
		End Property
		Public Property X2() As Integer
			Get
				Return FX2
			End Get
			Set(ByVal value As Integer)
				If FX2 <> value Then
					FX2 = value
					FModified = True
				End If
			End Set
		End Property
		Public Property Y1() As Integer
			Get
				Return FY1
			End Get
			Set(ByVal value As Integer)
				If FY1 <> value Then
					FY1 = value
					FModified = True
				End If
			End Set
		End Property
		Public Property Y2() As Integer
			Get
				Return FY2
			End Get
			Set(ByVal value As Integer)
				If FY2 <> value Then
					FY2 = value
					FModified = True
				End If
			End Set
		End Property
		#End Region

		Public ReadOnly Property Modified() As Boolean
			Get
				Return FModified
			End Get
		End Property
	End Class

	Public Class SelectionOperation
		Inherits TreeListOperation

		Private block As Block
'INSTANT VB NOTE: The field result was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private result_Renamed As String = String.Empty
		Private Const CellDelimeter As String = vbTab
		Private Const LineDelimeter As String = vbCrLf

		Public Sub New(ByVal block As Block)
			Me.block = block
		End Sub

		Public ReadOnly Property Result() As String
			Get
				Return result_Renamed
			End Get
		End Property

		Public Overrides Sub Execute(ByVal node As DevExpress.XtraTreeList.Nodes.TreeListNode)
			If Not block.Between(block.Y1, block.Y2, node.Id) Then
				Return
			End If
			For Each column As TreeListColumn In node.TreeList.Columns
				If block.Contains(column.VisibleIndex, node.Id) Then
					result_Renamed &= node.GetDisplayText(column)
					result_Renamed &= CellDelimeter
				End If
			Next column
			result_Renamed &= LineDelimeter
		End Sub
	End Class

End Namespace
